### PR TITLE
community: add reka chat model integration

### PR DESCRIPTION
**Pull request by findalexli**
community: add Reka chat model integration

Description: This PR adds support for the Reka chat model to the LangChain community package. It includes the implementation of ChatReka class, which allows users to interact with Reka's API for chat completions. The integration supports both text and image inputs, streaming responses, and async operations.

Dependencies: This integration inside langchain-community requires the reka-api package to be installed. Users will need to install it separately with pip install reka-api.

Added features:

ChatReka class for interacting with Reka chat models
Support for text and image inputs
Streaming and non-streaming response handling
Synchronous and asynchronous methods
Unit and integration tests
Example notebook demonstrating usage
Testing:

Unit tests have been added in tests/unit_tests/chat_models/test_reka.py
Integration tests have been added in tests/integration_tests/chat_models/test_reka.py
An example notebook has been created in docs/docs/integrations/chat/reka.ipynb, including tool use with search. 
I have run make format, make lint, and make test from the root of the community package, and all checks pass.

X/Twitter handle: @RekaAILabs

### CODE REVIEW

**Comment by efriis** [comment]
closing as duplicate of #26405 - let me know if they're different

**Comment by efriis** [comment]
nvm - got it - personal branch

**Comment by efriis** [comment]
looks like CI might still be failing though

**Comment by findalexli** [comment]
thanks. we pushed a few fix to address the CI issues.

 @efriis  we don't have the credentials to look into the vercel deployment issue, can you help with that?

File: libs/community/extended_testing_deps.txt [hunk 1]
```diff
@@ -82,7 +83,7 @@ sseclient-py>=1.8.0,<2
 streamlit>=1.18.0,<2
 sympy>=1.12,<2
 telethon>=1.28.5,<2
-tidb-vector>=0.0.3,<1.0.0
+tidb-vector>=0.0.3,<1.0.0
```
Comment (vbarda) [inline libs/community/extended_testing_deps.txt:86]:
```suggestion
tidb-vector>=0.0.3,<1.0.0
```

File: libs/community/langchain_community/chat_models/reka.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+    ToolMessage,
+)
+from langchain_core.outputs import ChatGeneration, ChatGenerationChunk, ChatResult
+from langchain_core.runnables import Runnable
+from langchain_core.tools import BaseTool
+from langchain_core.utils import get_from_dict_or_env
+from langchain_core.utils.function_calling import convert_to_openai_tool
+from pydantic import BaseModel, ConfigDict, Field, model_validator
+
+REKA_MODELS = [
```
Comment (vbarda) [inline libs/community/langchain_community/chat_models/reka.py:42]: this doesn't seem to be used anywhere

File: libs/community/langchain_community/chat_models/reka.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        Returns:
+            Runnable: An executable chain or component.
+        """
+        formatted_tools = [
+            convert_to_openai_tool(tool, strict=strict) for tool in tools
+        ]
+
+        # Ensure tool_choice is one of the allowed options
+        if tool_choice not in ("auto", "none", "tool"):
+            warnings.warn(
```
Comment (vbarda) [inline libs/community/langchain_community/chat_models/reka.py:392]: this should be an error instead

**Comment by findalexli** [comment]
Our model is not trained with system prompt. Our api combines them just like our langchain integration.

File: libs/community/tests/unit_tests/chat_models/test_reka.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                    "role": "assistant",
+                    "content": [
+                        {"type": "text", "text": "It's a beautiful landscape."}
+                    ],
+                },
+            ],
+        ),
+    ],
+)
+def test_convert_to_reka_messages(
```
Comment (vbarda) [inline libs/community/tests/unit_tests/chat_models/test_reka.py:126]: i think this needs a few more tests for working with system messages

**Comment by vbarda** [comment]
@findalexli thanks for the contribution! i added a few comments, apologies for incomplete review last week

File: libs/community/langchain_community/chat_models/reka.py [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+        elif isinstance(message, HumanMessage):
+            processed_content = process_content(message.content)
+            if system_message:
+                if (
+                    processed_content
+                    and isinstance(processed_content[0], dict)
+                    and processed_content[0].get("type") == "text"
+                    and "text" in processed_content[0]
+                ):
+                    processed_content[0]["text"] = (
```
[the remark this answers is not in the document]
Comment (vbarda) [reply libs/community/langchain_community/chat_models/reka.py:99]: > Our model is not trained with system prompt. Our api combines them just like our langchain integration.

saw your response, threading here. if the API already handles system messages and converts them, why do this in the LangChain integration as well?

**Comment by findalexli** [comment]
Addressed the tests and a couple issues you pointed out, thank you @vbarda !

**Comment by vbarda** [comment]
> Addressed the tests and a couple issues you pointed out, thank you @vbarda !

@findalexli thanks for adding more tests! there is still a bunch of comments that are not addressed, or marked resolved but don't seem to be resolved. could you please respond to those and update the PR where appropriate?

**Comment by findalexli** [comment]
HI @vbarda  we made some updates and have addressed all of your comments (including notebook, token count, duplicated class, system role).  Added a patch in the `test_import` that were the source failing the ci test. Run lint and format. Thanks!

File: libs/community/langchain_community/chat_models/reka.py [hunk 6]
```diff
@@ -152,8 +154,9 @@ class RekaCommon(BaseLanguageModel):
     default_request_timeout: Optional[float] = None
     max_retries: int = 2
     reka_api_key: Optional[str] = None
-    count_tokens: Optional[Callable[[str], int]] = None
     model_kwargs: Dict[str, Any] = Field(default_factory=dict)
+    model_config = ConfigDict(extra="forbid")
+    _tiktoken_encoder = None
```
Comment (vbarda) [inline libs/community/langchain_community/chat_models/reka.py:159]: why is this private? could we make it into a more general `token_counter` param with better types? something like this https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/messages/utils.py#L577

File: libs/community/langchain_community/chat_models/reka.py [hunk 7]
```diff
@@ -339,11 +331,9 @@ async def _agenerate(
 
     def get_num_tokens(self, text: str) -> int:
         """Calculate number of tokens."""
-        if self.count_tokens is None:
-            raise NotImplementedError(
-                "get_num_tokens() is not implemented for Reka models."
-            )
-        return self.count_tokens(text)
+        if self._tiktoken_encoder is None:
```
Comment (vbarda) [inline libs/community/langchain_community/chat_models/reka.py:334]: suggest making this into a more generic `token_counter` attribute that's not private

**Comment by vbarda** [comment]
> HI @vbarda we made some updates and have addressed all of your comments (including notebook, token count, duplicated class, system role). Added a patch in the `test_import` that were the source failing the ci test. Run lint and format. Thanks!

Thanks! Looks like there are a few things failing (due to missing `tiktoken` dependency)

**Comment by findalexli** [comment]
Updated to address the token count comment. thanks for the feedback. I have ran make lint, make lint and make test, and make extended_tests which passed.

File: libs/community/langchain_community/chat_models/reka.py [hunk 8]
```diff
@@ -329,11 +330,29 @@ async def _agenerate(
 
         return ChatResult(generations=[ChatGeneration(message=message)])
 
-    def get_num_tokens(self, text: str) -> int:
+    def get_num_tokens(self, input: Union[str, BaseMessage, List[BaseMessage]]) -> int:
```
Comment (vbarda) [inline libs/community/langchain_community/chat_models/reka.py:333]: looks like the input type here is not matching the one in the `token_counter` callable (failing linter)

**Review by vbarda — COMMENTED** [review]
looks good to me -- thanks for addressing! i think one lint issue remaining and should be good to merge

**Pull request merged**